api/events/src/main/java/io/opentelemetry/api/events/EventEmitter.java

```diff
@@ -32,6 +32,17 @@
 @ThreadSafe
 public interface EventEmitter {
 
+  /**
+   * Emit an event.
+   *
+   * @param epochNanos The time at which the event happened, in epoch nanoseconds.
+   * @param eventName the event name, which acts as a classifier for events. Within a particular
+   *     event domain, event name defines a particular class or type of event.
+   * @param attributes attributes associated with the event
+   */
+  @SuppressWarnings("InconsistentOverloads")
+  void emit(long epochNanos, String eventName, Attributes attributes);
```

Comment:
I always anticipated adding more fields. I think we should follow the pattern of [Logger](https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/logs/Logger.java) / [LogRecordBuilder](https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/logs/LogRecordBuilder.java).

Events are a little different though, since unlike log records, they have a couple of required fields in the domain, name, and body. Well, body is questionable, but inclined to keep it required to really hammer home the usage pattern.

I imagine an API like:
```
// Shorthand syntax
eventEmitter.emit("foo", Attributes.builder()...build());

// Builder syntax
eventEmitter.builder("foo", Attributes.builder()...build())
  .setTimestamp(...)
  .emit();
```

WDYT?

Reply:
Oh yeah interesting. I can take a swing at that.

Reply:
I guess I'm a little confused. So `eventEmitter.builder()` would return an `EventEmitterBuilder` instance, whose `build()` methods returns an `EventEmitter` instance, but now also gets a new `emit()` method that kinda implicitly calls `build().emit()`? This is getting cloudy.

I'm sure we've talked about this, but it trips me up that the `LogRecordBuilder` doesn't serve to create a `LogRecord` (eg builder has no build method). 

I think the way I look at it is that users just need a way to emit an event. Ideally, just one method somewhere. For the sake of discussion, let's just say something like

```
Events.emit(timestamp, domain, eventName, attributes, schemaUrl, instrumentationVersion)
```

but that's a little heavy, because most users won't change the schemaUrl or version or even domain very often, and usually they want "now", so we really want a curry'd version of this, more like

```
emitter = Events.curriedEmit(clock, domain, schemaUrl, instrumentationVersion)
emitter.emit(eventName, attributes)
emitter.emit(eventName, attributes)
```

We've packaged away the invariants and created some nicer ergonomics. We don't really have a curry, but to me the `EventEmitterBuilder` is kinda serving that purpose, so when you call build() you have a preconfigured emitter. The thing that's weird is that the timestamp is not an invariant. It's different for every event, whether it comes from a clock via `now()` or whether its passed it. It has to be obtained at runtime. 

So if we don't like the overloaded `emit()` on the interface, I think we have to fracture the types. Go Java.

This suggests two interfaces -- one that's a `NowClockEventEmitter` that only has `emit(name, attrs)` and a `TimestampedClockEventEmitter` that has `emit(time, name, attrs)`. I don't love that either and I feel like it's getting complicated (overengineered).

Maybe the approach is to have an `EventBuilder` that only has an `emit()` method. That seems much more similar to the `LogRecordBuilder` approach then. Just talking this out to get my thoughts down.

Reply:
> I guess I'm a little confused. So eventEmitter.builder() would return an EventEmitterBuilder instance, whose build() methods returns an EventEmitter instance, but now also gets a new emit() method that kinda implicitly calls build().emit()? This is getting cloudy.

No, EventEmitterBuilder builds EventEmitter. eventEmitter.builder(String, Attributes) would return an EventBuilder.

> Maybe the approach is to have an EventBuilder that only has an emit() method. That seems much more similar to the LogRecordBuilder approach then. Just talking this out to get my thoughts down.

Yes 🙂

Reply:
Ok, I made an attempt at making the thing described here. I don't love it. 

I think this results in entirely too much construction code and the mixing of builders that `emit()` and emitters with `builders()` is going to be confusing to the user. It does seem to be somewhat more consistent with other signals.

Reply:
Let's delete this method. If you want to set the timestamp, call `builder(String, Attributes).setTimeStamp().build()`. 

Will result in 1 shorthand method for the most common use case: `emit(String, Attributes)`

And a builder for any other case where more than just event name and attributes is required.

Reply:
Just bumping this in case this was lost in the shuffle.

Reply:
FYI I pushed a commit to remove this emit overload. Now the only path to setting the timestamp is through the builder. I also adjusted the setTimestamp to match the conventions elsewhere in this repo by having a setTimestamp(long, TimeUnit)` and `setTimestamp(Instant)` variants.